### DISCUSSION TITLE
Get rid of legacy temporal extract functions

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#4745] [Design Proposals]
I just realized that in dialects we still register the legacy temporal extract functions:

* `year`
* `month`
* `day`
* `hour`
* `minute`
* `second`
* `week`
* `quarter`

although in the parser we interpret uses like `second(..)` as `extract(second from ..)`, so the functions that dialects register through:

* `org.hibernate.dialect.function.CommonFunctionFactory#weekQuarter`
* `org.hibernate.dialect.function.CommonFunctionFactory#hourMinuteSecond`
* `org.hibernate.dialect.function.CommonFunctionFactory#yearMonthDay`

are never called. We should either stop interpreting `second(..)` as `extract(second from ..)` or remove the unnecessary function registrations.

Note that functions in databases like e.g. `second(..)` might have a different return type than `extract(second from ..)` which can lead to some surprises when users want to use the native function.

**Comment by sebersole** [comment]
I'm guessing that you have no concrete example of this, right?

However, I do agree that it is reasonable to use the database's specific function if it defines one (`second` in your example) rather than the standard `extract` function.

As for specifics, it is `SemanticQueryBuilder#visitDatetimeField` you are talking about?

**Reply by beikov** [reply]
>I'm guessing that you have no concrete example of this, right?

The `second` example was what I ran into. I had a query like `select second(d.createdDate) from Document d` and expected an `int` as return, but then I realized that it now returns `float` because it uses `extract` behind the scenes which defines `float` as type for the unit `second`.

>As for specifics, it is `SemanticQueryBuilder#visitDatetimeField` you are talking about?

Kind of, yeah. If you look at the current rule in the grammar:

```
extractFunction
	: EXTRACT LEFT_PAREN extractField FROM expression RIGHT_PAREN
	| datetimeField LEFT_PAREN expression RIGHT_PAREN
	;
```

and `SemanticQueryBuilder#visitExtractFunction` which handles this, you will see that it always makes use of `extract` in the end. I have a "fix" for this prepared locally, but wanted to start a discussion first.

**Reply by gavinking** [reply]
_[UPDATE] This is wrong._

@sebersole the `Dialect` registrations actually just delegate to the definition of `extract()` for that database. So Christian is right, we don't need them in both places.

I think we can remove this branch from the parser rule `| datetimeField LEFT_PAREN expression RIGHT_PAREN` and everything would just keep working.

**Reply by beikov** [reply]
> the Dialect registrations actually just delegate to the definition of extract() for that database.

Not sure where that happens. Can you point me to that maybe?

How about we use a function if there is one registered under the name, and if there is none, "emulate" through `extract`?

**Reply by gavinking** [reply]
_[UPDATE] This is wrong too._

Ah, sorry, I'm wrong. It's the *parser* that's responsible for desugaring `year(x)` to `extract(year from x)`, my bad, I misremembered.

And then `CommonFunctionFactory.yearMonthDay()` overrides that for databases which do have these functions natively.

OK, then I agree with Steve, this is good, I don't think we should change it.

**Reply by gavinking** [reply]
> I don't think we should change it.

I mean we *could* change it to work like I said: do the desugaring in the `Dialect.initializeFunctionRegistry()` instead of in the parser, and that would be fine, but it's actually more code and wouldn't change the behavior at all, so I'm not sure what the benefit would be.

**Reply by gavinking** [reply]
ooooh wait, I see what you're saying now: you're saying that those things registered in `Dialect.initializeFunctionRegistry()` are never called because `year()` and friends get desugared to `extract()` too early.

OK, sure, that's true. So there potentially *is* a benefit in desugaring them in `initializeFunctionRegistry()` instead of in the parser. Well, fine, I guess. I'm not against doing that.

**Reply by beikov** [reply]
>It's the parser that's responsible for desugaring `year(x)` to `extract(year from x)`

Due to this, the functions registered in the dialects are never used. So IMO we should either remove the function registrations or do the desugaring only if there is no explicit function registered.

**Reply by gavinking** [reply]
Or we could just leave it alone because it's fine how it is .... 🤷‍♂️

Don't care much.

**Reply by gavinking** [reply]
>  IMO we should either remove the function registrations

Please don't do that: they're useful documentation for the future about which functions which databases support. It took me a *lot* of work to compile all that information.

> do the desugaring only if there is no explicit function registered

Meh, we could do that. If you like more code instead of less code.

I dunno, I just see this as a nonissue. There's no bug here.

**Reply by beikov** [reply]
I already wrote that it might be confusing for users if they want to override these with custom functions because due to the desugaring, these custom functions will never be called. But the "bigger" problem is that, for example `second` returns `int` on some databases, whereas `extract(second)` returns a `float`.

>Meh, we could do that. If you like more code instead of less code.

It's a simple if block in `SemanticQueryBuilder`, so not a big deal.

>I dunno, I just see this as a nonissue. There's no bug here.

I ran into this, so I suppose others could run into that as well.

**Reply by gavinking** [reply]
>  But the "bigger" problem is that, for example `second` returns int on some databases, whereas `extract(second)` returns a float.

Ah, well, good point, I forgot that. In that case I've changed my mind again: _please don't change this._

Look, it's documented in both the new HQL docs, _and_ in the Javadoc for `initializeFunctionRegistry()` that these are synonyms (abbreviations) for `extract()`:

```java
	 * Finally, the following functions are defined as abbreviations
	 * for extract(), and desugared by the parser:
	 *
	 * <ul>
	 * <li> second(arg)					- synonym of extract(second from a)
	 * <li> minute(arg)					- synonym of extract(minute from a)
	 * <li> hour(arg)					- synonym of extract(hour from a)
	 * <li> day(arg)					- synonym of extract(day from a)
	 * <li> month(arg)					- synonym of extract(month from a)
	 * <li> year(arg)					- synonym of extract(year from a)
	 * </ul>
```

So if you make this change you're changing the documented semantics that I really did spend a lot of time thinking about. I didn't just go doing arbitrary things. Sure, I can't recall the details off the top of my head _today_, two years later, but please please please just stop trying to change shit I worked really hard on and put a *lot* of thought into at the time. Because I can't always remember the details, but I at least attempted to _document_ them, as you can see above.

**Reply by gavinking** [reply]
And in the documentation:

```AsciiDoc
The following synonyms for `extract()` are also provided:

|===
| `year(x)` | `extract(year from x)`
| `month(x)` | `extract(month from x)`
| `day(x)` | `extract(day from x)`
| `hour(x)` | `hour(year from x)`
| `minute(x)` | `minute(year from x)`
| `second(x)` | `second(year from x)`
|===
```

**Reply by gavinking** [reply]
> `extract(second)` returns a float.

_Bugger_ I fucked this up in the JPA spec god damn it all to hell:

https://github.com/eclipse-ee4j/jpa-api/pull/314/files

**Reply by gavinking** [reply]
I have proposed a fix https://github.com/eclipse-ee4j/jpa-api/pull/354, let's hope it gets accepted.

Now I'm going to go off and feel bad about myself for the rest of the day.

**Reply by gavinking** [reply]
God damn it it's not OK to make mistakes like this.

**Reply by sebersole** [reply]
@gavinking You've gone back and forth enough here that I'm not sure whether to say I agree or disagree ;)

From my point of view, it is surprising if -

1. My database supports a `seconds` function
2. I use the function `seconds` in HQL
3. Hibernate decides to use `extract` instead.

Yes... I get the argument about "consistent return type" across databases.  If that is what they really care about however, they'd just use `extract` directly.  Or we could do this "remapping" better and always coerce the values correctly at the SQL level.  E.g. here we might cast to INTEGER.

Remaining consistent with documentation that was literally just written is also not super convincing to me.

**Reply by gavinking** [reply]
The issue is of portability and consistent behavior. Databases are all over the place on this, and so we have to impose our own order.

Now, if people think that the `second()` function should really return integer seconds and not fractional seconds then fine, OK, now we're having a different discussion  about what the semantics *should* be. If you guys think that, then fine, document that change and implement it. I'm not convinced it would be an improvement but I really don't care.

Note that doing this would take some actual work since not all DBs have a `second()` function IIRC (though most do). _UPDATE: "most" is an overstatement._

And it would be a *functional* change that you would be making because you think it's somehow better for some reason. (Nothing in the preceding discussion was proposing a change to semantics rather than to the implementation.)

What I care about is not breaking the consistent *semantics* across databases.

**Comment by sebersole** [comment]
It's an unfortunate, albeit understandable, choice of function names when more than one database defines explicit functions with those names.  So then what are your words of advice for existing applications that want `seconds` to still mean `seconds`?

> The issue is of portability and consistent behavior. Databases are all over the place on this, and so we have to impose our own order.

That is simply using different words for exactly what I said.  "Yes" is perfectly acceptable and more succinct ;)

**Reply by gavinking** [reply]
> It's an unfortunate, albeit understandable, choice of function names when more than one database defines explicit functions with those names.

So I believe I'm now better-able to reconstruct my reasoning on this. So I know for sure that I started out with all these seven (`year`, `month`, `day`, `hour`, `minutes`, `seconds`, `microseconds`) as separate functions, defined independently of `extract()`. But, of course, on some databases&mdash;including PostgreSQL, Oracle, SQL Server, and Sybase, among others&mdash;they're not defined natively, and so they would need to be delegated right back to `extract()` on those databases.

And so I guess I just thought: well, this is a waste of time maintaining *two* abstractions&mdash;for functions that just do the _exact same thing_&mdash;and it's just easier to say they're synonyms. Less code, and less for the user to need to understand.

But sure, I admit I didn't focus very much on the possibility of breaking people who're relying on `second()` meaning integer seconds. And that's partly I guess because I have a lot of trouble imagining why they would be ....

> So then what are your words of advice for existing applications that want `second` to still mean seconds?

Well, see that's where honestly I'm just not sure what to say. I mean, as it stands right now `floor(second(x))` is pretty close to the same thing I guess.

But the issue is this: as I said, I really have trouble imagining what anyone uses `second()` or `extract(second ...)` for!

0. I have a pretty clear idea of what people use `year`, `month`, `quarter`, `week_of_month`, and `day_of_week` for. They use them for calendaring applications. Similarly, I can understand how `hour` and `minute` are used in that context. But `second`s just aren't used in calendaring.
1. You see, `time`s in principle come equipped with seconds, but in practice they never have them. There are no events which occur at `12:34:23` every Wednesday, for example! Yeah sure, sure, you might be able to construct some bizarre hypothetical, but let's be realistic here: in practice, _times don't have seconds_, not unless they're part of a timestamp. 
2. `timestamp`s do have seconds and fractional seconds&mdash;indeed, you probably want as many fractional seconds as possible in a `timestamp`&mdash;but with `timestamp`s you don't usually use `extract()`! You usually compare two timestamps with `>` or you subtract them, something like `x-y by second` or whatever. Extracting the fields of a timestamp doesn't seem right. Extracting fields is really meant for calendaring.
3. Now, sure, fine, you do sometimes *format* `Timestamp`s for display purposes, but using `extract()` or `second()` for that seems like just about the worst way to do it. We have other much nicer ways to format timestamps that are easier to use, and aware of i18n, etc.

So I just don't know how to reason about a use case that I don't really think exists.

So, finally: alright, if you're worried about the breakage to the `second()` function for existing programs, then fine. That's a perfectly legitimate thing to be worried about. We can separate these things. I'm neither for it nor against it.

And you're lucky in that I actually have most of the infrastructure already there in case we need to do that. (In particular, the `Dialect`s which *do* support these functions natively already register them. And the rest I guess I can just easily enough delegate to `extract()`.

So if that's what you want me to do, I can implement that pretty quickly. At the time I worked on all this, that felt like unnecessary complexity, but whatever, it's not something I have a strong opinion about.

So just tell me if that's what you want.

**Comment by sebersole** [comment]
To resolve this, I believe we should just use the form which gives consistency across all databases (aka the generalized `second` function as opposed to the Dialect-specific ones) and get rid of the Dialect registrations.

Unless I am mistaken, I think that is what y'all believe as well.

**Reply by beikov** [reply]
Yes, it just feels unnecessary to have the function registrations in code when these are never used.

**Reply by gavinking** [reply]
I think they're useful documentation of what functions the database supports.
